tellapart/aurproxy/mirror.py

```diff
@@ -0,0 +1,329 @@
+      command = self._render(self._template_path, context)
+    return command
+
+  def _generate_context(self):
+    """
+    Build context necessary to render gor command.
+
+    Returns:
+      Context dictionary.
+    """
+    context = {}
+    context['gor_path'] = self._gor_path
+    context['ports'] = self._ports
+    context['endpoints'] = self._source.endpoints
+    context['max_qps'] = self._max_qps
+    return context
+
+  def _render(self, template_path, context):
+    """
+    Render gor command.
+
+    Args:
+      template_path - str - path to gor command template.
+      context - dict - parameters to gor command template.
+
+    Returns:
+      Rendered gor command.
+    """
+    with open(template_path) as t:
+      template = Template(t.read())
+    return template.render(**context)
+
+  def _update(self, command, command_path, kill_running):
+    """
+    Update mirroring process dynamic launch script.
+
+    Args:
+      command - str - traffic mirroring process command.
+      command_path - path to process mirroring dynamic launch script.
+      kill_running - boolean - whether to kill running mirroring processes.
+
+    Returns:
+      Boolean indicating whether update was successful.
+    """
+    updated = self._update_command(command, command_path)
+    success = False
+    if updated and kill_running:
+      # Currently no graceful restart mechanism in gor
+      # Depend on Aurora to start gor back up
+      success = self._kill_running()
+    return updated and success
+
+  def _update_command(self, command, command_path):
+    """
+    Apply update of dynamic launch script.
+
+    Args:
+      command - str - traffic mirroring process command.
+      command_path - path to process mirroring dynamic launch script.
+
+    Returns:
+      Boolean indicating whether update was applied.
+    """
+    updated = False
+    # Don't rewrite it if nothing has changed.
+    if os.path.isfile(command_path):
+      with open(command_path) as cp:
+        if cp.read() == command:
+          logger.info('Mirror command is unchanged.')
+          updated = True
+          return updated
+
+    logger.info('Writing new mirror command.')
+    logger.info('Command: {0}'.format(command))
+    try:
+      with os.fdopen(os.open(command_path,
+                             os.O_WRONLY | os.O_CREAT | os.O_TRUNC,
+                             0755), 'w') as command_handle:
+        command_handle.write(command)
+        updated = True
+    except Exception:
+      logger.exception('Attempt to update mirror command failed!')
+    return updated
+
+  def _kill_running(self):
+    """
+    Kill running traffic mirror and fallback processes.
+
+    Returns:
+      Whether successful in killing all traffic mirror and fallback processes.
+    """
+    logger.info('Looking for existing traffic mirror processes.')
+    success = True
+    killed_any = False
+    try:
+      for proc in psutil.process_iter():
```

Comment:
Hm, could we use a pidfile or something similar here? This creates hard dependency on process isolation within Docker.